tutorials/ImplementASolver.qmd

```diff
@@ -0,0 +1,340 @@
+- `get_iterate(rws)` to access the current iterate
+
+We recommend to follow the general scheme with the `stop` field. If you have specific criteria
+when to stop, consider implementing your own [stoping criterion](https://manoptjl.org/stable/plans/stopping_criteria/) instead
+
+### Initialization & Iterate Access
+
+For our solver, there is not so much to initialize, just to be safe we should copy over the
+initial value in `p` we start with, to `q`. We do not have to care about remembering the iterate,
+that is done by `Manopt.jl`. For the iterate access we just have to pass `p`.
+
+```{julia}
+#| output: false
+function initialize_solver!(mp::AbstractManoptProblem, rws::RandomWalkState)
+    copyto!(M, rws.q, rws.p) # Set p^{(0)} = q
+    return rws
+end
+get_iterate(rws::RandomWalkState) = rws.p
+get_solver_result(rws::RandomWalkState) = rws.q
+```
+
+and siilarly we implement the step. Here we make use of the fact that the problem
+(and also the objective in fact) have access functions for their elements,
+the one we need is [`get_cost`](@ref).
+
+```{julia}
+#| output: false
+function step_solver!(mp::AbstractManoptProblem, rws::RandomWalkState, i)
+    M = get_manifold(mp) # for ease of use get the manifold from the problem
+    X = rand(M; vector_at=p)     # generate a direction
+    X .*= rws.σ/norm(M, p, X)
+    # Walk
+    retract!(M, rws.p, rws.p, X, rws.retraction_method)
+    # is the new point better? Then store it
+    if get_cost(mp, rws.p) < get_cost(mp, rws.q)
+        copyto!(M, rws.p, rws.q)
+    end
+    return rws
+end
+```
+
+Performance wise we could improve the number of allocations by making `X` also a field of
+our `rws` but let's keep it simple here.
+We could also store the cost of `q` in the state, but we will see how to easily also enable
+this solver to allow for [caching](https://manoptjl.org/stable/tutorials/CountAndCache/#How-to-Count-and-Cache-Function-Calls).
+
+Now we can just run the solver already! We take the same example as for the other tutorials
+
+We first define our task, the Riemannian Center of Mass from the [Get Started: Optimize!](https://manoptjl.org/stable/tutorials/Optimize!.html) tutorial.
+
+```{julia}
+#| output: false
+Random.seed!(23)
+n = 100
+σ = π / 8
+M = Sphere(2)
+p = 1 / sqrt(2) * [1.0, 0.0, 1.0]
+data = [exp(M, p,  σ * rand(M; vector_at=p)) for i in 1:n];
+f(M, p) = sum(1 / (2 * n) * distance.(Ref(M), Ref(p), data) .^ 2)
+```
+
+We can now generate the problem with its objective and the state
+
+```{julia}
+mp = DefaultManoptProblem(M, ManifoldCostObjective(f))
+s = RandomWalkState(M; σ = 0.2)
+
+solve!(mp, s)
+get_solver_result(s)
+```
+
+The function `solve!` works also in place of `s`, but the last line illustrates how to access
+the result in general; we could also just look at `s.p`, but the function `get_iterate` is
+also used in several other places.
+
+We could for example easilz set up a second solver to work from a specified starting point
```

Comment:
```suggestion
We could for example easily set up a second solver to work from a specified starting point
```

Reply:
Oh, yes, I wrote some par t of that at home on my German keyboard, and the one thing to get consistent then is the switch of z and y on said keyboard ;)